llvm/lib/MC/MCCASObjectV1.cpp

```diff
@@ -1082,13 +1116,17 @@ Error MCCASBuilder::createDebugStrSection() {
 Error MCCASBuilder::buildFragments() {
   startGroup();
 
+  Expected<SmallVector<DebugInfoCURef>> CURefs = splitDebugInfoSection();
+  if (!CURefs)
```

Comment:
I understand what you are doing here, but I have a different idea to consider. You are creating the individual compile unit cas objects outside the loop so that you can get the abbreviation offsets from them and then do the abbreviations, then in the loop you are just creating the debuginfo section cas block out of the array of the `CURefs`

The way I had envisioned it, and this is just a suggestion, is that you move the whole compile unit building out of the loop and just skip the iteration in the for loop for the debuginfo section.

This way you do not need to break the current `createDebugInfoSection` into 3 separate functions. You just have `createDebugInfoSection` return an `Expected<Smallvector<uint32_t>,0>> AbbrevOffsets` instead. How does that sound?

Reply:
>you move the whole compile unit building out of the loop and just skip the iteration in the for loop for the debuginfo section.

This would make the order of sections in the final object file different from the baseline object file, causing the verifier to fail. Does this make sense?

Reply:
While semantically the order doesn't matter (I think?), we probably want to generate identical binaries with and without the CAS, so we must respect the section order

Reply:
Yes you are correct, I didn't realize that